oauth2-server/src/main/java/com/clouway/oauth2/authorization/ClientAuthorizationRepository.java

```diff
@@ -2,38 +2,51 @@
 
 import com.clouway.oauth2.DateTime;
 import com.clouway.oauth2.client.Client;
+import com.clouway.oauth2.codechallenge.CodeChallenge;
 import com.google.common.base.Optional;
 
 import java.util.Set;
 
 /**
  * ClientAuthorizationRepository is a repository which is keeping records for the authroziations
  * that are performed for the client applications.
- * 
+ *
  * @author Ivan Stefanov <ivan.stefanov@clouway.com>
  */
 public interface ClientAuthorizationRepository {
 
   /**
    * Finds authorization that is associated with the provided authCode.
    *
-   * @param client the client for which authorization was issued
+   * @param client   the client for which authorization was issued
    * @param authCode the code of the authorization
-   * @param instant the time on which check is performed
+   * @param instant  the time on which check is performed
    * @return the authorization or absent value
    */
   Optional<Authorization> findAuthorization(Client client, String authCode, DateTime instant);
-
+//If we decide to use the new authorize method for all requests we could remove this as it is only used in the example app
+// Comment will be removed on code review
   /**
-   * Authorize authorizes the identity using
+   * Authorize authorizes the identity
    *
-   * @param client the client which is requesting authorization
-   * @param identityId the identity to be authorized
-   * @param scopes scopes for the authorization code
+   * @param client       the client which is requesting authorization
+   * @param identityId   the identity to be authorized
+   * @param scopes       scopes for the authorization code
    * @param responseType the type of the response to be returned
    * @return the authorization or absent value if identity was not authorized
    */
   Optional<Authorization> authorize(Client client, String identityId, Set<String> scopes, String responseType);
 
+  /**
+   * Authorize authorizes clients with an included codeChallenge for better security
+   *
+   * @param client        the client which is requesting authorization
+   * @param identityId    the identity to be authorized
+   * @param scopes        scopes for the authorization code
+   * @param responseType  the type of the response to be returned
+   * @param codeChallenge the code challenge sent by the client
+   * @return the authorization or absent value if the identity was not authorized
+   */
+  Optional<Authorization> authorize(Client client, String identityId, Set<String> scopes, String responseType, CodeChallenge codeChallenge);
```

Comment:
This method becomes good candidate for Dependency Inversion: 

```java
class CodeChallengeAuthorization implements ClientAuthorization {

  public CodeChallengeAuthorization(CodeChallenge codeChallenge) {
    // ...
  }
  public Optional<Authorization> authenticate(Client client, String identity, Set<String> scopes, String responseType){
    // ...     
  }
}

class BasicAuthorization implements ClientAuthorization {
    public Optional<Authorization> authenticate(Client client, String identity, Set<String> scopes, String responseType){
    // ...
  }
}
```

e.g the moving parts between implementations

Reply:
Could you elaborate more on the matter? I don't think i am understanding the idea correctly.

Reply:
It's not a good idea methods to be designed to accepts parameters which will be used only in some cases.

Reply:
We want to support both ways of authorization. It looks strange but i think that in the case when an authorization code is generated a blank CodeChallenge to be sent to the repository. A possible solution to avoid this would be to separate the CodeChallenge and not save it with the Authorization object in the ClientAuthorizationRepository. Still i think they should be kept together as they are tightly coupled.

Reply:
Introduction of 2 implementation will force re-design of the authorization store. 

It maybe hidden by introducing of client authorization request. It will not look explicit in the implementation 

```java
class ClientAuthorizationRequest {
   Client client;
   String  identity;
   String responseType;   
   CodeChallenge challenge;  //nullable type 
}

// definition: 
Optional<Authorization> authorize(ClientAuthorizationRequest request)
```

Reply:
btw, naming here is misleading: 
method says that it's authentication but returns authorization. I suppose method should be named authorize instead.

Reply:
Originally the method is called authorize yeah. This looks better as well. We will have to do a change in the implementation of ClientAuthorizationRepository anyway.

Reply:
Yes, challenge need to be persistent and later used for the code exchange call.